src/FwdState.cc

```diff
@@ -1268,6 +1288,28 @@ FwdState::exhaustedTries() const
     return n_tries >= Config.forward_max_tries;
 }
 
+bool
+FwdState::supportsRepinning() const
+{
+    assert(request->flags.pinned);
+
+    // pconn race on pinned connection: Currently we do not have any mechanism
+    // to re-pin current pinned connection path.
+    if (!serverDestinations.empty() && serverDestinations[0] == nullptr)
```

Comment:
What are the use cases where `request->flags.pinned` is true but `(!serverDestinations.empty() && serverDestinations[0] == nullptr)` is false?

Reply:
pconn(/pinned) race cases. In this case we should retry the same path (repin on the same path)

Reply:
> why would (!serverDestinations.empty() && serverDestinations[0] == nullptr) be false when request->flags.pinned is true?

Yes you are right. As is now this is can not be true. This is because the nil path from serverDestinations removed after this is checked.
My commit 13cff134 in practice disables re-pining in all cases, the pinned connections are never retried or re-forwarded.

The correct check is something like the  "request->flags.pinned && pconnRace == raceHappened".

Reply:
fixed at 4c206d0b